fix: added fuzzy search logic for datasource home

AmanAgarwal041 opened the pull request:
## Description
Earlier when we were searching `sk` , snowflake was not returned in the searched datasources. Hence after this pr instead of exacting matching a fuzzy search logic is applied which helps us in searching for snowflake datasource even if we add `sk` in the search input. 

Steps to test: 
- Go to application
- Go to Add new datasources page
- Add `sk` in the search input, snowflake should be  present in the UI.

Fixes #38983
_or_  
Fixes `Issue URL`
> [!WARNING]  
> _If no issue exists, please create an issue first, and check with the maintainers if the issue is valid._

## Automation

/ok-to-test tags="@tag.Datasource"

### :mag: Cypress test results
<!-- This is an auto-generated comment: Cypress test results  -->
> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/13172714737>
> Commit: ca099a3e1eca8c8a5953d508f9833d00fd94d7b1
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=13172714737&attempt=2" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.Datasource`
> Spec:
> <hr>Thu, 06 Feb 2025 06:58:30 UTC
<!-- end of auto-generated comment: Cypress test results  -->


## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [ ] No


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
  - Enhanced search filtering now delivers alphabetically sorted and more accurate results when browsing plugins and datasources.
  - Added a new prop for mock datasources to improve the functionality of the Empty Searched Plugins component.

- **Refactor**
  - Streamlined and centralized the search filtering process across integration pages for a more consistent and scalable user experience.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

AmanAgarwal041 commented:
/build-deploy-preview skip-tests=true

NilanshBansal commented:
@rahulbarwal can you please review this PR

rahulbarwal commented:
I am sorry I don't have context on this issue. @AmanAgarwal041, would it be possible for you to add some more details in the description about what the problem was, what's the new solution and how exactly to test this. I would like to test this on the DP first.

AmanAgarwal041 commented:
> I am sorry I don't have context on this issue. @AmanAgarwal041, would it be possible for you to add some more details in the description about what the problem was, what's the new solution and how exactly to test this. I would like to test this on the DP first.

Updated the description and added steps to test as well.

rahulbarwal reviewed (commented):
Overall I agree with the changes in the sense that it will accomplish your goal. But we already are using fuse.js in our repo that will be a better replacement than our own native implementation which is susceptible to edge cases failure. 

https://github.com/appsmithorg/appsmith/blob/release/app/client/src/selectors/templatesSelectors.tsx#L157

AmanAgarwal041 commented:
/build-deploy-preview skip-tests=true

rahulbarwal commented:
@AmanAgarwal041 On the DP when I search for `sl`, the result `MySQL` is not coming up.

![image](https://github.com/user-attachments/assets/49d27dd1-9149-40b6-866d-0f5179da2490)

AmanAgarwal041 commented:
/build-deploy-preview skip-tests=true

AmanAgarwal041 commented:
> @AmanAgarwal041 On the DP when I search for `sl`, the result `MySQL` is not coming up.
> 
> ![image](https://private-user-images.githubusercontent.com/6761673/410301606-49d27dd1-9149-40b6-866d-0f5179da2490.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg4MTk4MzAsIm5iZiI6MTczODgxOTUzMCwicGF0aCI6Ii82NzYxNjczLzQxMDMwMTYwNi00OWQyN2RkMS05MTQ5LTQwYjYtODY2ZC0wZjUxNzlkYTI0OTAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDIwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAyMDZUMDUyNTMwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjdhMjQ2YzI0Nzc1NjkyZTUyYzE2NTI2MTAzOWQwOTI3NjFkZTkzMzljMDBjYzI0MjNkZWU3NDdjMmExZmVlNCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.LTFHvlxTwolmnvLrJX-wcjUlTjsMnX3zZ6FGFdbIAM4)

@rahulbarwal Fixed the threshhold for the search

rahulbarwal commented on app/client/src/pages/Editor/IntegrationEditor/util.ts:
```diff
@@ -0,0 +1,20 @@
+  shouldSort: true,
+  threshold: 0.6,
+  location: 0,
+  distance: 100,
```
Nit: 100 distance seems too much for datasources name search, as we have limited options to search from. 

See if reducing this make sense.
